crates/optimism/rpc/src/eth/transaction.rs

```diff
@@ -29,9 +32,30 @@ where
         self.inner.signers()
     }
 
-    /// Decodes and recovers the transaction and submits it to the pool.
-    ///
-    /// Returns the hash of the transaction.
+    async fn transaction_by_hash(
+        &self,
+        hash: B256,
+    ) -> Result<Option<RpcTransaction<Self::NetworkTypes>>, Self::Error> {
+        Ok(LoadTransaction::transaction_by_hash(self, hash).await?.map(|tx| {
+            let is_deposit = tx.as_recovered().is_deposit();
+            let mut tx = tx.into_transaction::<Self::TransactionCompat>();
+            // deposit receipt version for given transaction, if the block number is known
+            if is_deposit && tx.block_number().is_some() {
+                tx.deposit_receipt_version = self
+                    .inner
+                    .provider()
+                    .chain_spec()
+                    .is_fork_active_at_timestamp(
+                        OptimismHardfork::Canyon,
+                        tx.block_number().unwrap_or(0),
+                    )
+                    .then_some(1);
+            }
+
+            tx
+        }))
```

Comment:
imo this is the only relevant change here, but this is incomplete because we also need to fetch the receipt here because we need to set the once to the receipt's deposit nonce unfortunately

Reply:
how is it set for receipt?

Reply:
this is how we set it here:
https://github.com/paradigmxyz/reth/blob/6dca2a5236e5a2c6317b27a5cdfb3d17dcd96dbc/crates/optimism/evm/src/strategy.rs#L218-L224

But using the block number for a timestamp is not correct. We should just fetch the receipt rather than rely on the tx meta + fork checks here.

Reply:
exactly! same as by @aroralanuk . then no need to load the whole receipt to get deposit receipt version at least, only chainspec needed, hence this pr isn't blocked and a good start

Reply:
Ya, that's where I took it from. Also, in my initial change, I had `is_fork_active_at_block` not `is_fork_active_at_timestamp` as tx has a block_number and no block_timestamp. Should i make this change again?

Reply:
No, that's sort of the issue, is that the rpc tx does not have block timestamp information, and the hardfork is a timestamp-activated hardfork. So using block number here will not work, even if you use `is_fork_active_at_block`

Reply:
Ohh I see. Can we fetch the timestamp from the block_header ie Header?

```rust
 let block_timestamp = self.inner.provider().header_by_number_or_tag(block_number).unwrap().timestamp;
 ```